File: 0033-search-in-rotated-sorted-array/0033-search-in-rotated-sorted-array.cpp
class Solution {
public:
    int binSearch(vector<int> &v, int k, int si, int ei){
        while(si <= ei){
            int mid = si+(ei-si)/2;
            if(v[mid] == k)return mid;
            else if(v[mid] < k)si = mid+1;
            else ei = mid-1;
        }
        return -1;
    }
    int search(vector<int>& nums, int target) {
        int pvtIdx = nums.size();

        int si = 1, ei = pvtIdx-1;

        while(si <= ei){
            int mid = si + (ei-si)/2;
            if(nums[mid] < nums[mid-1]){pvtIdx = mid;break;}
            if(nums[mid] >= nums[0])si = mid+1;
            else ei = mid-1;
        }

        if(target >= nums[0]){
            return binSearch(nums,target,0,pvtIdx-1);
        }
        else return binSearch(nums,target,pvtIdx,nums.size()-1);
    }
};

File: 0033-search-in-rotated-sorted-array/0033-search-in-rotated-sorted-array.java
class Solution {
    public int search(int[] nums, int target) {
        int n=nums.length;
        int ans=0;
        for(int i=1;i<n;i++){
            if(nums[i]<nums[i-1]){
               ans=i;
             break;
            }
        }
          int low1=0;
          int high1=ans-1;
          while(low1<=high1){
            int mid=(low1+high1)/2;
            if(nums[mid]==target)return mid;
             if(nums[mid]>target){
                high1=mid-1;
             }
             else{
                low1=mid+1;
             }

          }
          int low2=ans;
          int high2=n-1;
          while(low2<=high2){
            int mid=(low2+high2)/2;
            if(nums[mid]==target)return mid;
             if(nums[mid]>target){
                high2=mid-1;
             }
             else{
                low2=mid+1;
             }

          }
          return -1;
    }
}